File: src/ticket_service.ts
import * as restate from "@restatedev/restate-sdk";
import {
  protoMetadata,
  Ticket,
  TicketService,
} from "./generated/proto/example";
import { BoolValue } from "./generated/proto/google/protobuf/wrappers";
import { Empty } from "./generated/proto/google/protobuf/empty";

class TicketSvc implements TicketService {
  async markAsSold(request: Ticket): Promise<Empty> {
    return Promise.resolve(Empty.create({}));
  }

  async reserve(request: Ticket): Promise<BoolValue> {
    return Promise.resolve(BoolValue.create({}));
  }

  async unreserve(request: Ticket): Promise<Empty> {
    return Promise.resolve(BoolValue.create({}));
  }
}

restate.createServer().bindService({
  service: "TicketService",
  instance: new TicketSvc(),
  descriptor: protoMetadata,
}).listen(8082);

File: java/src/main/java/dev/restate/tour/app/TicketService.java
package dev.restate.tour.app;

import com.google.protobuf.BoolValue;
import dev.restate.sdk.RestateContext;
import dev.restate.sdk.common.TerminalException;
import dev.restate.tour.generated.TicketServiceRestate;
import dev.restate.tour.generated.Tour.Ticket;

public class TicketService extends TicketServiceRestate.TicketServiceRestateImplBase {
    @Override
    public BoolValue reserve(RestateContext ctx, Ticket request) throws TerminalException {
        return BoolValue.of(true);
    }

    @Override
    public void unreserve(RestateContext ctx, Ticket request) throws TerminalException {
    }

    @Override
    public void markAsSold(RestateContext ctx, Ticket request) throws TerminalException {
    }
}
